cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/GoalUtils.java

```diff
@@ -151,25 +153,41 @@ private static void filterOutBrokersExcludedForReplicaMove(List<Broker> original
 
   /**
    * Check whether the proposed action is legit. An action is legit if it is:
-   * (1) a replica movement and the destination does not have a replica of the same partition and is allowed to have a
-   * replica from the partition, or
-   * (2) a leadership movement, the replica is a leader and the destination broker has a follower of the same partition.
+   * (1) a replica movement across brokers, the destination broker does not have a replica of the same partition and is
+   * allowed to have a replica from the partition
+   * (2) a replica movement across the disks of the same broker, the destination disk is alive
+   * (3) a leadership movement, the replica is a leader and the destination broker has a follower of the same partition.
    *
    * @param replica Replica that is affected from the given action type.
    * @param destinationBroker Destination broker.
+   * @param destinationDisk Destination disk.
    * @param clusterModel Cluster model.
    * @param actionType Action type.
    * @return True if the move is legit, false otherwise.
    */
-  static boolean legitMove(Replica replica, Broker destinationBroker, ClusterModel clusterModel, ActionType actionType) {
-    if (actionType == ActionType.REPLICA_MOVEMENT
-        && clusterModel.partition(replica.topicPartition()).canAssignReplicaToBroker(destinationBroker)
-        && destinationBroker.replica(replica.topicPartition()) == null) {
-      return true;
+
+  static boolean legitMove(Replica replica,
```

Comment:
Rather than getting `destinationBroker` and `destinationDisk`, would `destinationDisk` suffice to infer the broker?